File: scala/ch01/ex1.32.scala
def accumulate[A](combiner: (A,A) => A, null_value: A, 
    term: (Int) => A, a: Int, next: (Int) => (Int), b: Int): A = {
  if (a > b) null_value else
      combiner(term(a), accumulate(combiner, null_value, term, next(a), next, b))
}

def accumulate2[A](combiner: (A,A) => A, null_value: A, 
    term: (Int) => A, a: Int, next: (Int) => (Int), b: Int) = {
  def ac(a2: Int, accum: A): A = if (a2 > b) accum else
      ac(next(a2), combiner(term(a2), accum))
  ac(a, null_value)
}

def prod(term: (Int) => Int, a: Int, next: (Int) => (Int), b: Int) = 
  accumulate((i:Int,j:Int) => i*j, 1, term, a, next, b)

def prod2(term: (Int) => Int, a: Int, next: (Int) => (Int), b: Int): Int = 
  accumulate2((i:Int,j:Int) => i*j, 1, term, a, next, b)

def sum(term: (Int) => Double, a: Int, next: (Int) => (Int), b: Int) = 
  accumulate((x:Double,y:Double) => x+y, 0.0, term, a, next, b)

def sum2(term: (Int) => Double, a: Int, next: (Int) => (Int), b: Int): Double = 
  accumulate2((x:Double,y:Double) => x+y, 0.0, term, a, next, b)

def even(n: Int) = n % 2 == 0
val cube = (n: Double) => n * n * n
def identity[A](n: A) = n
def inc(n: Int) = n + 1

def fact(n: Int)  = prod(identity, 1, inc, n)
def fact2(n: Int) = prod2(identity, 1, inc, n)

def simpsonsRule(f: (Double)=>Double, a: Double, b: Double, n: Int) = {
  val h = (b - a) / n
  def fy(k: Int) = f(a + (k * h))
  def term(i: Int) = fy(i) * (if (even(i)) 2 else 4)
  (h / 3) * (fy(0) + sum(term, 1, inc, (n - 1)) + fy(n))
}
def simpsonsRule2(f: (Double)=>Double, a: Double, b: Double, n: Int) = {
  val h = (b - a) / n
  def fy(k: Int) = f(a + (k * h))
  def term(i: Int) = fy(i) * (if (even(i)) 2 else 4)
  (h / 3) * (fy(0) + sum2(term, 1, inc, (n - 1)) + fy(n))
}

import org.scalatest._
import org.scalatest.matchers._

val factVals = Array(1, 2, 6, 24, 120, 720, 5040)
  
object factorialSpec extends Spec with ShouldMatchers {
  describe ("Factorials computed with recursive 'accumulate'") {
    it ("should calculate the correct values") {
      (1 to 7) foreach { n => fact(n) should equal (factVals(n-1)) }
    }
  }
  describe ("Factorials computed with iterative 'accumulate'") {
    it ("should calculate the correct values") {
      (1 to 7) foreach { n => fact2(n) should equal (factVals(n-1)) }
    }
  }
}

factorialSpec execute

object simpsonsRuleSpec extends Spec with ShouldMatchers {
  describe ("simpsonsRule computed with recursive 'accumulate'") {
    it ("should converge to the correct answer with more and more steps") {
      simpsonsRule(cube, 0.0, 1.0, 10)   should be (.25 plusOrMinus 0.01)
      simpsonsRule(cube, 0.0, 1.0, 100)  should be (.25 plusOrMinus 0.001)
      simpsonsRule(cube, 0.0, 1.0, 1000) should be (.25 plusOrMinus 0.0001)
      // Stack overflow!
      // simpsonsRule(cube, 0.0, 1.0, 10000) should be (.25 plusOrMinus 0.00001)
    }
  }
  describe ("simpsonsRule computed with iterative 'accumulate'") {
    it ("should converge to the correct answer with more and more steps") {
      simpsonsRule2(cube, 0.0, 1.0, 10)   should be (.25 plusOrMinus 0.01)
      simpsonsRule2(cube, 0.0, 1.0, 100)  should be (.25 plusOrMinus 0.001)
      simpsonsRule2(cube, 0.0, 1.0, 1000) should be (.25 plusOrMinus 0.0001)
      // No Stack overflow!
      simpsonsRule2(cube, 0.0, 1.0, 10000) should be (.25 plusOrMinus 0.00001)
    }
  }
}

simpsonsRuleSpec execute

File: clojure/ch01/ex1.32.clj
(ns sicp.ch01 (:use clojure.contrib.test-is))

(defn accumulate [combiner null_value term a next b]
  (if (> a b) null_value 
      (combiner (term a) (accumulate combiner null_value term (next a) next b))))

(defn accumulate2 [combiner null_value term a next b]
  (loop [a2 a accum null_value]
    (if (> a2 b) accum 
      (recur (next a2) (combiner (term a2) accum)))))

(defn prod  [term a next b] (accumulate  (fn [i j] (* i j)) 1 term a next b))
(defn prod2 [term a next b] (accumulate2 (fn [i j] (* i j)) 1 term a next b))

(defn sum  [term a next b] (accumulate  (fn [i j] (+ i j)) 0.0 term a next b))
(defn sum2 [term a next b] (accumulate2 (fn [i j] (+ i j)) 0.0 term a next b))

(defn ident [n] n)
(defn increment [n] (+ n 1))

(defn fact  [n] (prod  ident 1 increment n))
(defn fact2 [n] (prod2 ident 1 increment n))

(deftest test-fact
  (is (= 1    (fact 1)))
  (is (= 2    (fact 2)))
  (is (= 6    (fact 3)))
  (is (= 24   (fact 4)))  
  (is (= 120  (fact 5)))  
  (is (= 720  (fact 6)))
  (is (= 5040 (fact 7))))

(deftest test-fact2
  (is (= 1    (fact2 1)))
  (is (= 2    (fact2 2)))
  (is (= 6    (fact2 3)))
  (is (= 24   (fact2 4)))  
  (is (= 120  (fact2 5)))  
  (is (= 720  (fact2 6)))
  (is (= 5040 (fact2 7))))

(defn cube [n] (* n n n))
(defn abs [n] (if (> n 0) n (- n)))

(defn simpsons-rule [f a b n]
  (def h (/ (- b a) n))
  (defn fy [k] (f (+ a (* k h))))
  (defn term [i]
    (* (fy i) (if (even? i) 2 4)))
  (* (/ h 3) (+ (fy 0) (sum term 1 increment (- n 1)) (fy n))))

(defn simpsons-rule2 [f a b n]
  (def h (/ (- b a) n))
  (defn fy [k] (f (+ a (* k h))))
  (defn term [i]
    (* (fy i) (if (even? i) 2 4)))
  (* (/ h 3) (+ (fy 0) (sum2 term 1 increment (- n 1)) (fy n))))

(deftest test-simpsons-rule
  (is (> 0.001 (abs (- 0.25 (simpsons-rule cube 0.0 1.0 10)))))
  (is (> 0.001 (abs (- 0.25 (simpsons-rule cube 0.0 1.0 100)))))
  (is (> 0.001 (abs (- 0.25 (simpsons-rule cube 0.0 1.0 1000)))))
  ; Stack overflow
  ; (is (> 0.001 (abs (- 0.25 (simpsons-rule cube 0.0 1.0 10000)))))
  )

(deftest test-simpsons-rule2
  (is (> 0.001 (abs (- 0.25 (simpsons-rule2 cube 0.0 1.0 10)))))
  (is (> 0.001 (abs (- 0.25 (simpsons-rule2 cube 0.0 1.0 100)))))
  (is (> 0.001 (abs (- 0.25 (simpsons-rule2 cube 0.0 1.0 1000)))))
  (is (> 0.001 (abs (- 0.25 (simpsons-rule2 cube 0.0 1.0 10000))))))

(run-tests)

